src/query/pipeline/transforms/src/processors/transforms/sort/merger.rs

```diff
@@ -165,73 +165,57 @@ where
             return false;
         };
 
+        let input_index = cursor.input_index;
+        let start = cursor.row_index;
+
         let max_rows = self.limit.unwrap_or(self.batch_rows).min(self.batch_rows);
-        let (whole_block, next_cursor) = if self.sorted_cursors.len() == 1 {
-            (true, None)
+        let count = if self.sorted_cursors.len() == 1 {
+            (cursor.num_rows() - start).min(max_rows - self.temp_sorted_num_rows)
+        } else if !A::SHOULD_PEEK_TOP2 {
+            debug_assert!(!cursor.is_finished());
```

Comment:
In this code block, we continue to find the smallest row with equal.

So I think both heap and loser tree can share this branch. It's nothin with `SHOULD_PEEK_TOP2`

Reply:
Where is the top2 cursor now?

Reply:
I think it's also right to advance he heap to find the equal element.

Reply:
Consider the following example:
cursor_top_1:  [1,1,2,3 ...]
cursor_top_2:  [2,...]

for !SHOULD_PEEK_TOP2 output [1,1]
for SHOULD_PEEK_TOP2 output [1,1,2] 

The heap branch is not worse.